c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -2878,10 +2899,17 @@ class DeviceCachingAllocator {
         block->device,
         context ? context : block->context_when_segment_allocated);
 
-    C10_CUDA_CHECK(cudaFree((void*)block->ptr));
+    auto* pool = block->pool;
+    auto active_pool = MemPoolContext::getActiveMemPool();
+    if (active_pool && active_pool->allocator() && pool->owner_PrivatePool) {
+      // If there is an active mempool with a given allocator,
+      // we use the given allocator's delete function.
+      active_pool->allocator()->raw_delete((void*)block->ptr);
```

Comment:
I'm a little confused here: how do you guarantee that pool and active_pool are the same?

Reply:
The top level function that ends up calling this function (`release_block`) is `release_cached_blocks`. In `release_cached_blocks`, I ensure that only the blocks of those pools are released whose id matches that of the active pool:
```
    for (auto it = graph_pools_freeable.begin();
         it != graph_pools_freeable.end();) {
      if (mempool_id.first != 0 || mempool_id.second != 0) {
        if (it->first == mempool_id) {
          // If there is an active mempool, we sync only the events
          // associated with the pool
          synchronize_and_free_events(context, it->second);
        } else {
          // otherwise we move on
          ++it;
          continue;
        }
      }
      // See notifyCaptureDestroy for the strategy here.
      TORCH_INTERNAL_ASSERT(it->second->use_count == 0);
      release_blocks(it->second->small_blocks, context);
      release_blocks(it->second->large_blocks, context);
      if (it->second->cudaMalloc_count == 0) {
        auto erase_count = graph_pools.erase(it->first);
        TORCH_INTERNAL_ASSERT(erase_count == 1);
        it = graph_pools_freeable.erase(it);
      } else {
        ++it;
      }
    }
```
I they don't match, I `continue` out of the loop and the pool remains unreleased in `graph_pools_freeable`.

Reply:
OK. Consider adding an assert they match then.